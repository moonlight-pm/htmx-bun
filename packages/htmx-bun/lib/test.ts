import { afterAll } from "bun:test";
import { mkdtempSync, rmdirSync } from "node:fs";
import { tmpdir } from "node:os";

export function makeTemporaryDirectory() {
    const dir = mkdtempSync(`${tmpdir()}/htmx-bun-`);
    afterAll(() => {
        rmdirSync(dir, { recursive: true });
    });
    return dir;
}
