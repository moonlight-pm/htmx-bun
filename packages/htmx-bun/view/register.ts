import { Glob } from "bun";
import { resolve } from "path";
import { error, info, warn } from "~/lib/log";
import { Template, TemplateModule } from "./template";

export class TemplateRegister {
    templates: Record<string, Template> = {};

    constructor(
        // Base path for the view tree
        public base: string,
    ) {}

    async initialize() {
        for await (const path of new Glob("**/*.part").scan(this.base)) {
            const tag = tagFromPath(path);
            if (this.templates[tag]) {
                warn(
                    "view",
                    `Duplicate tag '${tag}' defined in '${this.base}/${path}'`,
                );
                warn(
                    "view",
                    `Using '${this.base}/${this.templates[tag].path}'`,
                );
                continue;
            }
            info("register", `registering partial at '${path}'`);
            await this.reload(path);
        }
    }

    async reload(path: string) {
        const tag = tagFromPath(path);
        const absoluteFilePath = resolve(`${this.base}/${path}`);
        if (this.templates[tag]) {
            delete require.cache[absoluteFilePath];
        }
        try {
            const module = (await import(absoluteFilePath)) as TemplateModule;
            this.templates[tag] = new Template(this, tag, path, module);
        } catch (e) {
            error("view", `Failed to load '${path}'`);
            // @ts-ignore
            const cause = e.cause?.toString();
            if (cause) {
                console.log(cause);
            }
            return;
        }
    }

    get(tag: string): Template | undefined {
        return this.templates[tag];
    }
}

function tagFromPath(path: string) {
    let tag = path
        .split("/")
        .filter((it) => !/^\(.*\)$/.test(it))
        .join("-")
        .replace(/\.part/g, "");
    if (tag === "index") {
        tag = "root";
    }
    if (tag.endsWith("-index")) {
        tag = tag.replace(/-index$/, "");
    }
    return tag;
}
