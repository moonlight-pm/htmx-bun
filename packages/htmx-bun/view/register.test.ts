import { expect, test } from "bun:test";
import { unlinkSync } from "fs";
import { TemplateRegister } from "./register";

test("initialize", async () => {
    const register = new TemplateRegister("./view/__fixtures__");
    await register.initialize();
    expect(register.get("root")).toBeDefined();
});

test("reload", async () => {
    const testPath = "./view/__fixtures__/reload.part";
    const testContent1 = "<div>Reload Test 1</div>\n";
    const testContent2 = "<div>Reload Test 2</div>\n";
    await Bun.write(testPath, testContent1);
    const register = new TemplateRegister("./view/__fixtures__");
    await register.initialize();
    expect(register.get("reload").html).toMatchSnapshot();
    await Bun.write(testPath, testContent2);
    await register.reload("reload");
    expect(register.get("reload").html).toMatchSnapshot();
    unlinkSync(testPath);
});
