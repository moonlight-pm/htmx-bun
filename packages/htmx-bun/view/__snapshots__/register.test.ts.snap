// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`reload 1`] = `
"<div>
  Reload Test 1
</div>
"
`;

exports[`reload 2`] = `
"<div>
  Reload Test 1
</div>
"
`;
